File: atsa/src/main/java/jp/kaiz/atsassistmod/ifttt/IFTTTUtil.java
package jp.kaiz.atsassistmod.ifttt;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.zip.GZIPInputStream;
import java.util.zip.GZIPOutputStream;

/** Serialises IFTTT rule containers (gzip + Jackson polymorphic), faithful port. */
public final class IFTTTUtil {
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private IFTTTUtil() {}

    public static byte[] toBytes(IFTTTContainer container) {
        try {
            return compress(MAPPER.writeValueAsBytes(container));
        } catch (IOException e) {
            return null;
        }
    }

    public static IFTTTContainer fromBytes(byte[] bytes) {
        try {
            return MAPPER.readValue(decompress(bytes), IFTTTContainer.class);
        } catch (IOException e) {
            return null;
        }
    }

    private static byte[] compress(byte[] bytes) throws IOException {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        try (GZIPOutputStream gzip = new GZIPOutputStream(out)) {
            gzip.write(bytes);
        }
        return out.toByteArray();
    }

    private static byte[] decompress(byte[] compressed) throws IOException {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        try (GZIPInputStream gzip = new GZIPInputStream(new ByteArrayInputStream(compressed))) {
            byte[] buffer = new byte[1024];
            int len;
            while ((len = gzip.read(buffer)) > 0) {
                out.write(buffer, 0, len);
            }
        }
        return out.toByteArray();
    }
}

File: atsa/src/main/kotlin/jp/kaiz/atsassistmod/ifttt/IFTTTUtil.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package jp.kaiz.atsassistmod.ifttt

import com.fasterxml.jackson.databind.ObjectMapper
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.util.zip.GZIPInputStream
import java.util.zip.GZIPOutputStream

/** Serialises IFTTT rule containers (gzip + Jackson polymorphic). */
object IFTTTUtil {
    private val mapper = ObjectMapper()

    @JvmStatic
    fun toBytes(container: IFTTTContainer): ByteArray? =
        try {
            compress(mapper.writeValueAsBytes(container))
        } catch (_: IOException) {
            null
        }

    @JvmStatic
    fun fromBytes(bytes: ByteArray): IFTTTContainer? =
        try {
            mapper.readValue(decompress(bytes), IFTTTContainer::class.java)
        } catch (_: IOException) {
            null
        }

    @Throws(IOException::class)
    private fun compress(bytes: ByteArray): ByteArray {
        val out = ByteArrayOutputStream()
        GZIPOutputStream(out).use { gzip ->
            gzip.write(bytes)
        }
        return out.toByteArray()
    }

    @Throws(IOException::class)
    private fun decompress(compressed: ByteArray): ByteArray {
        val out = ByteArrayOutputStream()
        GZIPInputStream(ByteArrayInputStream(compressed)).use { gzip ->
            val buffer = ByteArray(1024)
            while (true) {
                val length = gzip.read(buffer)
                if (length <= 0) {
                    break
                }
                out.write(buffer, 0, length)
            }
        }
        return out.toByteArray()
    }
}
